Add support for MAX16150/MAX16169

mrcsosa opened the pull request:
## PR Description

- Please replace this comment with a summary of your changes, and add any context
necessary to understand them. List any dependencies required for this change.
- To check the checkboxes below, insert a 'x' between square brackets (without
any space), or simply check them after publishing the PR.
- If you changes include a breaking change, please specify dependent PRs in the
description and try to push all related PRs simultaneously.

## PR Type
- [ ] Bug fix (a change that fixes an issue)
- [x] New feature (a change that adds new functionality)
- [ ] Breaking change (a change that affects other repos or cause CIs to fail)

## PR Checklist
- [x] I have conducted a self-review of my own code changes
- [ ] I have tested the changes on the relevant hardware
- [ ] I have updated the documentation outside this repo accordingly (if there is the case)

mrcsosa committed:
dt-bindings: input: add adi,max16150.yaml

Add documentation for device tree bindings for MAX16150/MAX16169

Signed-off-by: Marc Paolo Sosa <marcpaolo.sosa@analog.com>

mrcsosa committed:
input: misc: add driver for max16150

MAX16150/MAX16169 nanoPower Pushbutton On/Off Controller

Signed-off-by: Marc Paolo Sosa <marcpaolo.sosa@analog.com>

nunojsa commented on Documentation/devicetree/bindings/input/adi,max16150.yaml:
```diff
@@ -0,0 +1,62 @@
+            brcm,function = <0>;
+            brcm,pull = <0>;
+        };
+    };
```
This looks to be very rpi dependent... Typically not needed in bindings AFAIK. But you're definetly missing input properties. Where do you define the key code?

mrcsosa replied:
I think the key code is typically used to detect button presses. However, with MAX16150, it sends interrupt signals when certain amount of time the button is pressed.

nunojsa commented on Documentation/devicetree/bindings/input/adi,max16150.yaml:
```diff
@@ -0,0 +1,62 @@
+        compatible = "adi,max16150";
+        adi,variant = "A";
+
+        gpio-pb_in = <&gpio 17 1>;
```
Don't think the above needs to be a gpio. It's a button that you should register with the input subsystem

nunojsa commented on Documentation/devicetree/bindings/input/adi,max16150.yaml:
```diff
@@ -0,0 +1,62 @@
+        adi,variant = "A";
+
+        gpio-pb_in = <&gpio 17 1>;
+        gpio-out = <&gpio 18 0>;
```
I'm also not convinced we need the above...

nunojsa commented on Documentation/devicetree/bindings/input/adi,max16150.yaml:
```diff
@@ -0,0 +1,62 @@
+        gpio-pb_in = <&gpio 17 1>;
+        gpio-out = <&gpio 18 0>;
+        gpio-clr = <&gpio 22 1>;
+        gpio-int = <&gpio 23 0>;
```
This is an `interrupts` property and not a gpio

nunojsa commented on drivers/input/misc/max16150.c:
```diff
@@ -0,0 +1,227 @@
+
+	hrtimer_cancel(&data->debounce_timer);
+	hrtimer_cancel(&data->shutdown_timer);
+	return 0;
```
This can be done in a devm_action... Also not sure if we really need hrtimers for this

nunojsa commented on drivers/input/misc/max16150.c:
```diff
@@ -0,0 +1,227 @@
+	if (ret) {
+		dev_err(&pdev->dev, "Failed to request IRQ for PB_IN\n");
+		return ret;
+	}
```
Just use an irq... No need for it to be a gpio

nunojsa commented on drivers/input/misc/max16150.c:
```diff
@@ -0,0 +1,227 @@
+	} else {
+		dev_err(&pdev->dev, "Unknown device type: %s\n", type);
+		return -EINVAL;
+	}
```
use proper chip info structure for subtleties in the chips... The above does not scale at al.

nunojsa commented on drivers/input/misc/max16150.c:
```diff
@@ -0,0 +1,227 @@
+		return -EINVAL;
+	}
+
+	ret = device_property_read_string(&pdev->dev, "adi,variant", &variant);
```
Not in the bindings and I dunno you need a property for this... You already have compatibles and you can different info structure per compatible.

nunojsa reviewed (commented):
I'm stopping the review for now. Note that you pretty much copied the driver from one subsystem to another without any integration with the input subsystem. Not what was meant.

Please do not rush into opening PRs or having things done. Really try to understand what was asked and seek help if needed :wink:

kister-jimenez commented on drivers/input/misc/max16150.c:
```diff
@@ -0,0 +1,227 @@
+	}
+
+	switch (variant[0]) {
+	case 'A':
+		data->variant = MAX161X_A;
+		data->debounce_time = ktime_set(0, 50 * NSEC_PER_MSEC);
+		data->shutdown_time = (data->type == MAX16150) ? ktime_set(8,
+				      0) : ktime_set(16, 0);
+		break;
+	case 'B':
+		data->variant = MAX161X_B;
+		data->debounce_time = (data->type == MAX16150) ? ktime_set(2,
+				      0) : ktime_set(50, 0);
+		data->shutdown_time = ktime_set(16, 0);
+		break;
+	case 'C':
+		data->variant = MAX161X_C;
+		data->debounce_time = ktime_set(0, 50 * NSEC_PER_MSEC);
+		data->shutdown_time = ktime_set(16, 0);
+		break;
```
Not sure why you need the debounce time and shutdown time. 
Debounce time is for the push button input which should be connected to a push button not the processor system.

Shutdown period is how long the button is pressed before the chip deasserts (in some variant) and pull down the interrupt pin. This is also related to the button side.

kister-jimenez commented on drivers/input/misc/max16150.c:
```diff
@@ -0,0 +1,227 @@
+	dev_info(&pdev->dev, "Detected %s variant %s\n",
+		 (data->type == MAX16150) ? "MAX16150" : "MAX16169", variant);
+
+	data->gpio_pb_in = devm_gpiod_get(&pdev->dev, "pb_in", GPIOD_IN);
```
pb_in is not an input to the MPU and is not connected to the MPU. This is connected to the push button.

kister-jimenez commented on drivers/input/misc/max16150.c:
```diff
@@ -0,0 +1,227 @@
+	if (IS_ERR(data->gpio_pb_in))
+		return PTR_ERR(data->gpio_pb_in);
+
+	data->gpio_out = devm_gpiod_get(&pdev->dev, "out", GPIOD_OUT_LOW);
```
Out pin of the device is typically used to control a switch or to gate the power to the device. You don't need this.

kister-jimenez commented on drivers/input/misc/max16150.c:
```diff
@@ -0,0 +1,227 @@
+	if (IS_ERR(data->gpio_clr))
+		return PTR_ERR(data->gpio_clr);
+
+	data->gpio_int = devm_gpiod_get(&pdev->dev, "int", GPIOD_IN);
```
Just use an irq.

kister-jimenez commented on drivers/input/misc/max16150.c:
```diff
@@ -0,0 +1,227 @@
+#include <linux/ktime.h>
+#include <linux/of.h>
+
+enum max161x_variant {
+	MAX161X_A,
+	MAX161X_B,
+	MAX161X_C,
+};
+
+enum max161x_type {
+	MAX16150,
+	MAX16169,
+};
```
Why don't you just combine these? I think the variant is part of the ordering options for customers to select different timing options?

kister-jimenez commented on drivers/input/misc/max16150.c:
```diff
@@ -0,0 +1,227 @@
+{
+	struct max161x_data *data = dev_id;
+
+	if (gpiod_get_value(data->gpio_pb_in)) {
+		/* Button pressed */
+		hrtimer_start(&data->debounce_timer, data->debounce_time,
+			      HRTIMER_MODE_REL);
+	} else {
+		/* Button released */
+		if (data->variant == MAX161X_A && data->out_asserted) {
+			dev_info(data->dev, "Cancelling shutdown timer.");
+			hrtimer_cancel(&data->shutdown_timer);
+		}
+	}
```
You should use the input subsystem like input_report_key and input_sync to report the key and the value to the input subsystem and to sync input events.

I think what you want is to know the interval between the last irq if you want to distinguish the long press and the short momentary press.

kister-jimenez commented on drivers/input/misc/max16150.c:
```diff
@@ -0,0 +1,227 @@
+	hrtimer_init(&data->shutdown_timer, CLOCK_MONOTONIC, HRTIMER_MODE_REL);
+	data->shutdown_timer.function = max161x_shutdown_timer_cb;
+
+	ret = devm_request_irq(&pdev->dev, gpiod_to_irq(data->gpio_pb_in),
```
The irq pin is the int pin not the pb_in pin.

kister-jimenez reviewed (commented):
You might want to check out existing input device drivers for reference.

mrcsosa commented:
Changelogs:

-Converted few functions to use input subsystem
-Removed unused gpio pins, leaving int pin (to detect 32ms and 128ms interrupts) and clr pin to reset the output for Variant A devices.
-Removed variants on the driver code and dts
-

nunojsa commented on drivers/input/misc/max16150.c:
```diff
@@ -0,0 +1,117 @@
+		.of_match_table = max16150_of_match,
+	},
+};
+
```
nit: no need for new line in these cases

kister-jimenez commented on drivers/input/misc/max16150.c:
```diff
@@ -0,0 +1,117 @@
+			input_report_key(button, KEY_WAKEUP, 1);
+		input_sync(button);
+		falling_edge = false;
+	}
```
I agree with this. The datasheet also mentions this: 
`to signal the system to perform a specific function, or to initiate a
shutdown process` for the longer int. It also make sense to have the shorter int to have an option for a different event to be triggered. So, having 2 device properties for these 2, should be considered.

nunojsa replied:
Not sure it needs two events... I think the datasheet wording of "initiate a shutdown process" is just to state that the OUT pin will be deasserted. If it is a shutdown, ultimately depends on your plataform and where you connect that ouptut. I think we just have one event and assert/deassert states for it.

kister-jimenez commented on drivers/input/misc/max16150.c:
```diff
@@ -0,0 +1,119 @@
+	button->name = "max16150";
+	button->phys = "max16150/input0";
+	button->id.bustype = BUS_HOST;
+	input_set_capability(button, EV_KEY, KEY_POWER);
+	input_set_capability(button, EV_KEY, KEY_WAKEUP);
```
Ditto.

kister-jimenez commented on drivers/input/misc/max16150.c:
```diff
@@ -0,0 +1,111 @@
+	irq_number = gpio_to_irq(gpio_int);
+	ret = devm_request_irq(&pdev->dev, irq_number, max16150_isr,
+			       IRQF_TRIGGER_RISING, "max16150_irq", button);
+	if (ret)
```
I think both flags are needed here. That is, if we need to measure the falling to rising duration. Unless there's another way for this.

mrcsosa commented:
V3 changelogs:

Removed:
Remove function; max16150_remove.

Fixes:
Fixed code formatting, removed extra lines and white spaces.

Added:
added support for max16169 as I thought it won't be needed since it can apparently support the MAX16169 even the compatible device is only MAX16150.

nunojsa commented:
> V3 changelogs:

> Removed:
Remove function; max16150_remove.

> Fixes:
Fixed code formatting, removed extra lines and white spaces.

> Added:
added support for max16169 as I thought it won't be needed since it can apparently support the MAX16169 even the compatible device is only MAX16150.

I'm still seeing lots of comments I made in the previous review ignored. Either let me know why do you think those comments are not relevant so I can agree or disagree. At this point, not going to review again just to make the same comments :)

mrcsosa commented:
Hi, apologies for overseeing few of the comments, in this version I tried capturing and applying the comments I may have overseen.

### V4 Changelogs:

### New Additions:
**Custom Key Event Support:**
- Added a new variable `custom_key_event` initialized to `KEY_POWER`.
- Added code to read a custom key event from the device tree (`of_property_read_u32`).

**Interrupt Handling:**
- Split the interrupt handling into two separate functions: `max16150_rise_irq` and `max16150_fall_irq`.
- Added `IRQF_ONESHOT` flag to the interrupt request.

**Duration Calculation:**
- Introduced `irq_rising_time`, `irq_falling_time`, and `irq_low_duration` to calculate the duration between rising and falling edges.
- Added a threshold (`DURATION_THRESHOLD`) to determine the type of key event to report.

**Logging:**
- Added a `printk` statement to log the interrupt low duration.

### Changes:
**Removed Debounce Logic:**
-  Removed the debounce logic and the use of `delayed_work`.

**Simplified IRQ Handling:**
- Simplified the IRQ handling by directly calculating the duration and reporting the key event based on the duration.

**Removed GPIO Handling:**
- Removed `gpiod_get_value` and related GPIO handling functions.

**Removed `ktime` Usage:**
- Replaced `ktime` with `jiffies` for time calculations.

**Removed `input_unregister_device` in `remove` Function:**
- This version does not explicitly unregister the input device in the `remove` function. Remove function no longer available.

**Removed `devm_kzalloc` and `devm_gpiod_get`:**
- This does not allocate memory for the `max16150` structure or get the GPIO descriptor.

nunojsa commented on drivers/input/misc/max16150.c:
```diff
@@ -0,0 +1,149 @@
+	err = devm_request_any_context_irq(&pdev->dev, fall_irq,
+					   max16150_fall_irq,
+					   IRQF_TRIGGER_FALLING | IRQF_ONESHOT,
+					   "max16150_falling", bttn);
```
use normal threaded API

nunojsa commented on drivers/input/misc/max16150.c:
```diff
@@ -0,0 +1,149 @@
+
+	rise_irq = platform_get_irq(pdev, 1);
+	if (rise_irq < 0)
+		return rise_irq;
```
This is odd having two irqs for the same pin. But if it works, it is a good way to distinguish between rising and falling edge...

nunojsa commented on drivers/input/misc/max16150.c:
```diff
@@ -0,0 +1,149 @@
+
+	irq_rising_time = jiffies;
+
+	irq_low_duration = irq_rising_time - irq_falling_time;
```
this is odd like this... Just use normal ktime API... combine ktime_get() with ktime_to_ms())

nunojsa commented on drivers/input/misc/max16150.c:
```diff
@@ -0,0 +1,149 @@
+	if (irq_low_duration >= DURATION_THRESHOLD)
+		input_report_key(bttn, custom_key_event, 1);
+	else
+		input_report_key(bttn, KEY_POWER, 1);
```
you missed my point. We just have one key which by default is KEY_POWER (if not DT property is given) but that can be overwritten in DT

nunojsa commented on drivers/input/misc/max16150.c:
```diff
@@ -0,0 +1,149 @@
+	input_sync(bttn);
+
+	return IRQ_HANDLED;
+}
```
While this might be a good idea there's one thing that is still not perfect... If for some reason, the HW designers if one board where this device is placed remembers to invert the IRQ signal (because... why not :)), this whole logic will fail... Did you tried my suggestion with 

https://elixir.bootlin.com/linux/v6.6.65/source/include/linux/interrupt.h#L504

??

mrcsosa commented:
### Changelogs:

### max16150.c 
* Replaced jiffies to ktime for the duration checking.
* Removed the extra key event function along with the 128ms interrupt duration checking.
### adi,max16150.yaml
* Fixed the description for the linux,code and interrupts.
* Added include bindings irq.h and linux-event-codes.h
### Kconfig
* Changed the configuration and now uses tristate and changed the help text using other references in the input subsystem documentation.

nunojsa commented on Documentation/devicetree/bindings/input/adi,max16150.yaml:
```diff
@@ -0,0 +1,53 @@
+      Specifies a single numeric keycode value to be used for reporting
+      button/switch events. Specify KEY_RESERVED (0) to opt out of event
+      reporting.
+    $ref: /schemas/types.yaml#/definitions/uint32
```
no... please look at other examples using this. Hint: `git grep "linux,code" Documentation/devicetree`. 

The property is already defined so no need to do it again. You do need $ref the input bindings

nunojsa commented on Documentation/devicetree/bindings/input/adi,max16150.yaml:
```diff
@@ -0,0 +1,53 @@
+
+required:
+  - compatible
+  - linux,code
```
The code is not required as we the default is KEY_POWER. OTOH the interrupt is indeed a required property

nunojsa commented on drivers/input/misc/max16150.c:
```diff
@@ -0,0 +1,149 @@
+	cancel_delayed_work_sync(&button->debounce_work);
+	input_unregister_device(button->input);
+	return 0;
+}
```
Please do not just ignore comments... this .remove() callback is still here

nunojsa commented on drivers/input/misc/max16150.c:
```diff
@@ -0,0 +1,159 @@
+
+struct max16150_data {
+	struct input_dev *bttn;
+	int custom_key_event;
```
do we need the above?

nunojsa commented on drivers/input/misc/max16150.c:
```diff
@@ -0,0 +1,159 @@
+		return err;
+	}
+
+	platform_set_drvdata(pdev, data);
```
once you remove the .remove() callback, you do not need the above

nunojsa commented on drivers/input/misc/max16150.c:
```diff
@@ -0,0 +1,159 @@
+			data->custom_key_event = key_event;
+		else
+			dev_warn(&pdev->dev, "Failed to read custom key event, using default\n");
+	}
```
I already mentioned this... do not use OF but device properties. Please ask to me or someone else if you do not understand what i'm suggesting

nunojsa commented on drivers/input/misc/max16150.c:
```diff
@@ -0,0 +1,159 @@
+{
+	struct max16150 *button = platform_get_drvdata(pdev);
+
+	cancel_delayed_work_sync(&button->debounce_work);
```
the debounce_work is not a property of your struct

nunojsa commented on drivers/input/misc/max16150.c:
```diff
@@ -0,0 +1,159 @@
+					IRQF_TRIGGER_FALLING | IRQF_ONESHOT,
+					"max16150_falling", data);
+	if (err < 0) {
+		dev_err(&pdev->dev, "Can't register fall irq: %d\n", err);
```
`dev_err_probe()`. ditto for all the other places. You also need to state in the DT binbings that you need two interrupts

nunojsa replied:
please make sure to address previous comments...

nunojsa commented on drivers/input/misc/max16150.c:
```diff
@@ -0,0 +1,159 @@
+	}
+
+	data->bttn = bttn;
+	data->custom_key_event = KEY_POWER;
```
I would drip "custom" and just call it key_event or key_code

nunojsa commented on drivers/input/misc/max16150.c:
```diff
@@ -0,0 +1,159 @@
+
+	data->irq_falling_time = ktime_get();
+
+	input_report_key(bttn, KEY_POWER, 0);
```
not KEY_POWER.... data->key_event

nunojsa commented on drivers/input/misc/max16150.c:
```diff
@@ -0,0 +1,159 @@
+					     data->irq_falling_time);
+	unsigned long duration_ms = ktime_to_ms(irq_low_duration);
+
+	dev_info(dev, "Interrupt low duration: %lu ms\n", duration_ms);
```
drop the print... if you really want use dev_dbg_ratelimited()

nunojsa commented on drivers/input/misc/max16150.c:
```diff
@@ -0,0 +1,159 @@
+
+	input_report_key(bttn, data->custom_key_event, 1);
+	input_sync(bttn);
+
```
Where is the support for the required gpio on the chip that do not automatically de-asserts the output?

kister-jimenez replied:
Can the CLR GPIO just use the gpio-poweroff? So that whenever a power off command or event is triggered, the CLR pin is pulled low.

mrcsosa replied:
for an example, I tried using a separate script where a gpio is set to normally high, the same script captures the events send by the driver code, when long press is detected, it will toggle the gpio pin to low and deassert the clr pin

nunojsa commented on drivers/input/misc/max16150.c:
```diff
@@ -0,0 +1,159 @@
+
+	ktime_t irq_low_duration = ktime_sub(data->irq_rising_time,
+					     data->irq_falling_time);
+	unsigned long duration_ms = ktime_to_ms(irq_low_duration);
```
please no mixed code declaration...

nunojsa commented on drivers/input/misc/max16150.c:
```diff
@@ -0,0 +1,149 @@
+#include <linux/module.h>
+#include <linux/platform_device.h>
+#include <linux/jiffies.h>
+#include <linux/of.h>
```
you need mod_devicetable.h for the of_device_id table

nunojsa reviewed (commented):
Please do not rush into sending new version. Try to address all the comments and ask to me or a colleague if you fail to understand something. It will make both our lives easier as you won't need to spin this many revisions

mrcsosa committed:
Kconfig.adi imply MAX16150

Add entry for the MAX16150/MAX16169 driver.

Signed-off-by: Marc Paolo Sosa <marcpaolo.sosa@analog.com>

mrcsosa commented:
Changelogs
Enhancements & Feature Additions

 Short and Long Press Detection
- Added detection logic for short and long button presses in `max16150_fall_irq`.
- Defined press duration thresholds:
  - `SHORT_PRESS_MIN_MS = 32ms`
  - `SHORT_PRESS_MAX_MS = 127ms`
  - `LONG_PRESS_MIN_MS = 128ms`
- Short press logs `"Short press detected"` and sends `MSC_SCAN, 1`.
- Long press logs `"Long press detected"` and sends `MSC_SCAN, 2`.

 Input Subsystem Enhancements
- Added support for `EV_MSC` event with `MSC_SCAN` in `max16150_probe`.
- Now capable of distinguishing between short and long press events at the input level.

 Bug Fixes & Optimizations

IRQ Duration Calculation Moved
- Duration calculation for IRQ moved from `max16150_rise_irq` to `max16150_fall_irq`, ensuring accurate measurement of button press time.

Removed Redundant Logging
- Removed `dev_info(dev, "Button press detected")` from `max16150_rise_irq`, as press event logging is now handled in `max16150_fall_irq`.

General Code Cleanup
- Removed unnecessary `dev_dbg_ratelimited` logging for rising IRQ.
- Removed the remove function.

I still think that the gpio clr can be declared outside the driver.

nunojsa commented on Documentation/devicetree/bindings/input/adi,max16150.yaml:
```diff
@@ -0,0 +1,46 @@
+properties:
+  compatible:
+    description: |
+      Specifies the supported device variants. The MAX16150 and MAX16169 are supported.
```
Don't think the text needs to be formatted... This means you can drop '|'

nunojsa commented on Documentation/devicetree/bindings/input/adi,max16150.yaml:
```diff
@@ -0,0 +1,46 @@
+  interrupts:
+    maxItems: 1
+
+  linux,code: true
```
you need to $ref the proper input.yaml file to get this property

nunojsa commented on drivers/input/misc/max16150.c:
```diff
@@ -0,0 +1,154 @@
+#include <linux/platform_device.h>
+#include <linux/ktime.h>
+#include <linux/of.h>
+#include <linux/mod_devicetable.h>
```
not in alphabetical order and you still have of.h

nunojsa commented on drivers/input/misc/max16150.c:
```diff
@@ -0,0 +1,154 @@
+
+#define SHORT_PRESS_MIN_MS 32
+#define SHORT_PRESS_MAX_MS 127
+#define LONG_PRESS_MIN_MS 128
```
typically one uses the driver name as prefix... Example: MAX16150_SHORT_PRESS_MIN_MS

nunojsa commented on drivers/input/misc/max16150.c:
```diff
@@ -0,0 +1,154 @@
+		return err;
+	}
+
+	platform_set_drvdata(pdev, data);
```
not needed...

nunojsa commented on drivers/input/misc/max16150.c:
```diff
@@ -0,0 +1,154 @@
+	err = input_register_device(bttn);
+	if (err) {
+		dev_err(&pdev->dev, "Can't register input device: %d\n", err);
+		return err;
```
dev_err_probe()

nunojsa commented on drivers/input/misc/max16150.c:
```diff
@@ -0,0 +1,154 @@
+
+	data->bttn = bttn;
+	if (of_property_read_u32(np, "key-event", &data->key_event))
+		data->key_event = KEY_POWER;
```
Please use device property API. Fairly sure I already asked for it...

nunojsa commented on Documentation/devicetree/bindings/input/adi,max16150.yaml:
```diff
@@ -0,0 +1,46 @@
+    power-button {
+        compatible = "adi,max16150";
+        interrupt-parent = <&gpio>;
+        interrupts = <17 IRQ_TYPE_EDGE_BOTH>;
```
I doubt the driver works if DT is like this... Did you tested it?

nunojsa commented on drivers/input/misc/max16150.c:
```diff
@@ -0,0 +1,154 @@
+	err = devm_request_threaded_irq(&pdev->dev, fall_irq, NULL,
+					max16150_fall_irq,
+					IRQF_TRIGGER_FALLING | IRQF_ONESHOT,
+					"max16150_falling", data);
```
Thinking about this? Did you actually tested this? How did you described this in your DT? You would need to connect the interrupt signal into two different inputs in the processor. So this hack could be useful to know if you are in a rising or falling edge but it's not coherent with the bindings.

nunojsa commented on drivers/input/misc/max16150.c:
```diff
@@ -0,0 +1,154 @@
+	bttn->phys = "max16150/input0";
+	bttn->id.bustype = BUS_HOST;
+	input_set_capability(bttn, EV_KEY, data->key_event);
+	input_set_capability(bttn, EV_MSC, MSC_SCAN);
```
Why do you need the above?

nunojsa commented on drivers/input/misc/max16150.c:
```diff
@@ -0,0 +1,154 @@
+
+	data->irq_falling_time = ktime_get();
+	irq_duration = ktime_sub(data->irq_falling_time, data->irq_rising_time);
+	duration_ms = ktime_to_ms(irq_duration);
```
Well, can you actually ensure that the rising edge already happened?  I think you can still have a race during probe. So you could have requested the interrupt while your signal is already high and at that point, I do not think your max16150_rise_irq will trigger. Hence data->irq_rising_time will be 0 and you could wrongly detect a long press.

I think you could simple check data->irq_rising_time for 0 and ignore this interrupt in that case as there's no way to properly get the press duration. 

But this hack requesting two interrupts is very questionable... I'll ask this again: Did you tried my suggestion with this API?

https://elixir.bootlin.com/linux/v6.6.65/source/include/linux/interrupt.h#L504

If it works, you could properly do:

```
err = devm_request_threaded_irq(&pdev->dev, irq, NULL,
		max16150_irq,
		IRQF_TRIGGER_FALLING | IRQF_TRIGGER_RISING | IRQF_ONESHOT,
		"max16150_irq", data);
```

mrcsosa commented:
V5:

- Used dev_err_probe();
- Added chip_info
- removed interrupt and used gpio to track rising and falling edge better
- added gpio for clr pin
- added compatible variants A, B, and C
- removed time prints

mrcsosa commented on drivers/input/misc/max16150.c:
```diff
@@ -0,0 +1,159 @@
+				 "clr GPIO undefined, clr pin will not be asserted\n");
+		} else {
+			gpiod_set_value(max16150->clr_gpiod, 0);
+		}
```
We are using the clr gpio to reset the device, this is what we have implemented. I think the gpio should be high on boot since it will only go low if we want to reset or deassert the output.
```	if (chip_info->has_clr_gpio) {
		max16150->clr_gpiod = devm_gpiod_get(dev, "clr", GPIOD_OUT_HIGH);
		if (IS_ERR(max16150->clr_gpiod))
			return dev_err_probe(dev, PTR_ERR(max16150->clr_gpiod),
					     "Failed to get clr GPIO\n");

		if (max16150->clr_gpiod) {
			fsleep(1000);
			gpiod_set_value(max16150->clr_gpiod, 0);
		}
	}
```

nunojsa replied:
> We are using the clr gpio to reset the device, this is what we have implemented. I think the gpio should be high on boot since it will only go low if we want to reset or deassert the output.

Ahh, I get it know... But now that I thin about this, we should do this in another way because as we have it know, we will never use this pin for chips where has_clr_gpio = false. Sure, for those devices, it's not a mandatory pin but we should still be able to use it if we want too right? So, I would suggest something like:

```
max16150->clr_gpiod = devm_gpiod_get_optional(dev, "clr", GPIOD_OUT_HIGH);
if (IS_ERR(max16150->clr_gpiod))
    return dev_err_probe(dev, PTR_ERR(max16150->clr_gpiod),
           "Failed to get clr GPIO\n");
if (!max16150->clr_gpiod && chip_info->needs_clr_gpio)
    return dev_err_probe(dev, ...); //clr gpio is mandatory for latching out the output 
if (max16150->clr_gpiod) {
    ...
}
```

Makes sense?

nunojsa commented on Documentation/devicetree/bindings/input/adi,max16150.yaml:
```diff
@@ -0,0 +1,56 @@
+  interrupt-gpio:
+    maxItems: 1
+
+  clr-gpio:
```
you should make this mandatory for the compatible it makes sense. See examples here on how to do conditionals on bindings:

https://elixir.bootlin.com/linux/v6.13.4/source/Documentation/devicetree/bindings/iio/imu/adi,adis16475.yaml#L94

nunojsa replied:
Also it should be clr-gpios

nunojsa commented on Documentation/devicetree/bindings/input/adi,max16150.yaml:
```diff
@@ -0,0 +1,56 @@
+      - adi,max16169a
+      - adi,max16169b
+
+  interrupt-gpio:
```
should be interrupt-gpios

nunojsa commented on drivers/input/misc/max16150.c:
```diff
@@ -0,0 +1,158 @@
+		if (max16150->duration > MAX16150_LONG_INTERRUPT) {
+			if (max16150->chip_info->has_clr_gpio &&
+			    max16150->clr_gpiod)
+				gpiod_set_value(max16150->clr_gpiod, 1);
```
remove the if() and just do `gpiod_set_value(max16150->clr_gpiod, 1);`. The gpio subsystem is expected to handle the case where the gpiod is NULL (i.e do nothing)

nunojsa reviewed (commented):
Alright, I would say to take my last inputs and send this upstream

mrcsosa commented:
Changelogs:

Removed if() statement on probe for setting gpio, now directly handled by the gpio subsystem.
Minor changes at yaml file and added clr-gpios as mandatory requirement.

Thanks!!!

nunojsa commented on drivers/input/misc/max16150.c:
```diff
@@ -0,0 +1,161 @@
+// SPDX-License-Identifier: GPL-2.0-or-later
```
Drop the or-later

nunojsa commented on drivers/input/misc/max16150.c:
```diff
@@ -0,0 +1,161 @@
+/*
+ * Analog Devices MAX16150/MAX16169 Pushbutton Driver
+ *
+ * Copyright 2025 Analog Devices Inc.
```
2026 already :)

nunojsa commented on drivers/input/misc/max16150.c:
```diff
@@ -0,0 +1,161 @@
+#include <linux/input.h>
+#include <linux/interrupt.h>
+#include <linux/gpio/consumer.h>
+#include <linux/kernel.h>
```
The above brings a lot of unnecessary headers. Remove it and include only what you need.

nunojsa commented on drivers/input/misc/max16150.c:
```diff
@@ -0,0 +1,161 @@
+			input_report_key(max16150->input, max16150->keycode, 1);
+			input_sync(max16150->input);
+			input_report_key(max16150->input, max16150->keycode, 0);
+			input_sync(max16150->input);
```
So we report both power off and power up on the same "event"? Seems suspicious

nunojsa commented on drivers/input/misc/max16150.c:
```diff
@@ -0,0 +1,161 @@
+
+		if (!max16150->clr_gpiod)
+			return dev_err_probe(dev, -ENODEV,
+						 "clr GPIO is mandatory\n");
```
The above cannot happen given you're calling devm_gpiod_get()

nunojsa commented on drivers/input/misc/max16150.c:
```diff
@@ -0,0 +1,161 @@
+			return dev_err_probe(dev, -ENODEV,
+						 "clr GPIO is mandatory\n");
+
+		if (max16150->clr_gpiod) {
```
Just a note. The above if() is also redundant!

nunojsa reviewed (commented):
I think the only point of confusion to me is reporting key events! We should different conditions for different events. As the code stands and if I'm not missing nothing we report power off followed by power on (or the other way around) which is odd
